core/HoistComponent.ts

```diff
@@ -313,11 +315,16 @@ function wrapWithModel(render: RenderFn, cfg: Config): RenderFn {
         const managedRender = () => {
             let ctx = localModelContext;
             try {
+                const {testId} = props;
                 props.model = model;
                 delete props.modelRef;
                 delete props.modelConfig;
                 ctx.props = props;
                 ctx.modelLookup = newLookup ?? modelLookup;
+                if (testId && model) {
+                    instanceManager.registerModelWithTestId(testId, model);
```

Comment:
I think you are going to be registering the same thing over and over as testId and model are passed down through the hierarchy.  And it seems like the resolved model could be different for some intermediate components that are passing through the same test id?  

Also violating the "rules of hooks" that say you should *always* call the hook for any given component regardless of the state -- think may be ok, but only if we can guarantee that the result of that conditional to never change -- not sure we can. Let's discuss?

Reply:
First point, definitely true. My thought was that ultimately the deepest component with a given `testId` would register itself last, and that would end up being the associated model.  There would be several unnecessary calls to `instanceManager.registerModelWithTestId`, but that seemed like it could be ok given how lightweight it is?  Or maybe it’s something we only enable when in a test environment?  Would love to discuss all of this further.

Second point, I did consider this and thought it would be ok since the presence of `testId` *should* be immutable, though you could make an argument that it is a prop like any other and therefore doesn’t have to be (but that would sure be confusing!). Same thing with `model`- a component that has a model for one render *should* always have a model, just maybe not the same model.  Again could see this not being a guarantee.

Would you suggest moving the cleanup hook to the `finally` block so it always runs?  That said, if we’d be doing that to guard against a `testId` becoming `nil`, the hook would be ineffective anyway unless we cache the first `testId` which seems maybe overkill?  Would love to discuss!